libqtile/widget/redshift.py

```diff
@@ -0,0 +1,333 @@
+# Copyright (c) 2024 Saath Satheeshkumar(saths008)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import subprocess
+
+from libqtile.command.base import expose_command
+from libqtile.log_utils import logger
+from libqtile.widget import base
+from libqtile.widget.base import _TextBox
+
+
+class GammaGroup:
+    """
+    GammaGroup is a helper class to group redshift gamma settings.
+    """
+
+    def __init__(self, red, green, blue):
+        self.gamma_red = red
+        self.gamma_green = green
+        self.gamma_blue = blue
+
+    def _format_gamma(self, format_txt: str) -> str:
+        return format_txt.format(
+            gamma_red=self.gamma_red,
+            gamma_green=self.gamma_green,
+            gamma_blue=self.gamma_blue,
+        )
+
+    def _redshift_fmt(self) -> str:
+        return f"{self.gamma_red}:{self.gamma_green}:{self.gamma_blue}"
+
+    def __repr__(self) -> str:
+        return f"Gamma: {self._redshift_fmt()}"
+
+    def __str__(self) -> str:
+        return (
+            f"GammaGroup(red={self.gamma_red}, green={self.gamma_green}, blue={self.gamma_blue})"
+        )
+
+
+class RedshiftDriver:
```

Comment:
I don't see any reason for this to be a separate class. There's no preservation of any class variables etc. I think you could just include these methods in the widget itself.

Reply:
The main reason I did this was to make mocking easier (redshift isn't available in the test env so I could just mock out `redshift_driver`), is this reasonable or is there a better way to go about this?

Reply:
I thought that might be the reason. You can probably just mock the `subprocess.call` method instead. There are probably some examples of that in the test suite already.

Reply:
Done, removed.